// @evefrontier/dapp-kit@0.1.5/utils/constants.ts
// ============================================================================
// Environment Variable Helpers
// ============================================================================

/**
 * Get a required environment variable, throwing if not set.
 * @param name - The environment variable name (e.g., "VITE_SUI_GRAPHQL_ENDPOINT")
 * @throws {Error} If the environment variable is not set
 */
function getEnvVar(name: string): string {
  const value = import.meta.env[name];
  if (!value) {
    throw new Error(
      `Missing required environment variable: ${name}. ` +
        `Please set it in your .env file.`,
    );
  }
  return value;
}

// ============================================================================
// Environment-Based Configuration
// ============================================================================

/**
 * Get the Sui GraphQL endpoint URL from environment.
 * @returns The GraphQL endpoint URL
 * @throws {Error} If VITE_SUI_GRAPHQL_ENDPOINT is not set
 * @category Utilities
 */
export const getSuiGraphqlEndpoint = (): string =>
  getEnvVar("VITE_SUI_GRAPHQL_ENDPOINT");

/**
 * Get the EVE World package ID from environment.
 * @returns The package ID (0x-prefixed address)
 * @throws {Error} If VITE_EVE_WORLD_PACKAGE_ID is not set
 * @category Utilities
 */
export const getEveWorldPackageId = (): string =>
  getEnvVar("VITE_EVE_WORLD_PACKAGE_ID");

/**
 * Get the Character OwnerCap type string (derived from package ID).
 * @returns The fully qualified type string for Character OwnerCap
 * @category Utilities
 */
export const getCharacterOwnerCapType = (): string => {
  const pkg = getEveWorldPackageId();
  return `${pkg}::access::OwnerCap<${pkg}::character::Character>`;
};

/**
 * Get the Character PlayerProfile type string (derived from package ID).
 * @returns The fully qualified type string for Character PlayerProfile
 * @category Utilities
 */
export const getCharacterPlayerProfileType = (): string => {
  const pkg = getEveWorldPackageId();
  return `${pkg}::character::PlayerProfile`;
};

/**
 * Get the ObjectRegistry type string (derived from package ID).
 * @returns The fully qualified type string for ObjectRegistry
 * @category Utilities
 */
export const getObjectRegistryType = (): string =>
  `${getEveWorldPackageId()}::object_registry::ObjectRegistry`;

/**
 * Get the EnergyConfig type string (derived from package ID).
 * @returns The fully qualified type string for Energy Config
 * @category Utilities
 */
export const getEnergyConfigType = (): string =>
  `${getEveWorldPackageId()}::energy::EnergyConfig`;

/**
 * Get the Fuel Efficiency Config type string (derived from package ID).
 * @returns The fully qualified type string for Fuel Efficiency Config
 * @category Utilities
 */
export const getFuelEfficiencyConfigType = (): string =>
  `${getEveWorldPackageId()}::fuel::FuelConfig`;

/** Polling interval in milliseconds (10 seconds).
 *  @category Constants
 */
export const POLLING_INTERVAL = 10000;

/** Local storage keys.
 *  @category Constants
 */
export const STORAGE_KEYS = {
  CONNECTED: "eve-dapp-connected",
} as const;

/** Type IDs for in-game items.
 *  @category Constants
 */
/* eslint-disable @typescript-eslint/no-loss-of-precision */
export enum TYPEIDS {
  LENS = 77518,
  TRANSACTION_CHIP = 79193,
  COMMON_ORE = 77800,
  METAL_RICH_ORE = 77810,
  SMART_STORAGE_UNIT = 77917,
  PROTOCOL_DEPOT = 85249,
  GATEKEEPER = 83907,
  SALT = 83839,
  NETWORK_NODE = 88092,
  PORTABLE_REFINERY = 87161,
  PORTABLE_PRINTER = 87162,
  PORTABLE_STORAGE = 87566,
  REFUGE = 87160,
}

/** @category Constants */
export const EXCLUDED_TYPEIDS = [
  TYPEIDS.PORTABLE_REFINERY,
  TYPEIDS.PORTABLE_PRINTER,
  TYPEIDS.PORTABLE_STORAGE,
  TYPEIDS.REFUGE,
];

/** Volume - from wei units to m3.
 *  @category Constants
 */
export const ONE_M3 = 1000000000000000000;
